{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Immprint\n",
    "\n",
    "\n",
    "### Example\n",
    "\n",
    "Command-line tool:\n",
    "```\n",
    "immprint sample1a.csv sample1b.csv\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import immprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\r\n",
      "  File \"/home/thomas/.local/bin/immprint\", line 5, in <module>\r\n",
      "    from immprint.immprint import main\r\n",
      "  File \"/home/thomas/.local/lib/python3.8/site-packages/immprint/immprint.py\", line 35, in <module>\r\n",
      "    local_directory = inspect.getfile(olga)\r\n",
      "NameError: name 'olga' is not defined\r\n"
     ]
    }
   ],
   "source": [
    "!immprint sample1b.csv sample1a.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(immprint)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/usr/bin/env python3\r\n",
      "import argparse\r\n",
      "import numpy as np\r\n",
      "import inspect\r\n",
      "from pathlib import Path\r\n",
      "import olga\r\n",
      "import olga.load_model\r\n",
      "import olga.generation_probability\r\n",
      "import mpmath\r\n",
      "import gmpy2\r\n",
      "\r\n",
      "translate_header = {\r\n",
      "    'sequenceStatus': 'frame_type',\r\n",
      "    'productive': 'frame_type',\r\n",
      "    'aminoAcid': 'amino_acid',\r\n",
      "    'vGeneName': 'v_gene',\r\n",
      "    'v_call': 'v_gene',\r\n",
      "    'jGeneName': 'j_gene',\r\n",
      "    'j_call': 'j_gene',\r\n",
      "    'rearrangement': 'rearrangement',\r\n",
      "    'aminoAcid': 'cdr3_aa',\r\n",
      "    'cdr3_aa': 'cdr3_aa',\r\n",
      "    'cdr3s_aa': 'cdr3_aa',\r\n",
      "    'cdr3': 'cdr3_nt',\r\n",
      "    'cdr3s_nt': 'cdr3_nt',\r\n",
      "    'junction_aa': 'cdr3_aa',\r\n",
      "    'CDR3.amino.acid.sequence': 'cdr3_aa',\r\n",
      "    'CDR3.nucleotide.sequence': 'cdr3_nt',\r\n",
      "    'bestVGene': 'v_gene',\r\n",
      "    'bestJGene': 'j_gene',\r\n",
      "    'consensus_count': 'read_count',\r\n",
      "    'Read.count': 'read_count',\r\n",
      "    'reads': 'read_count'}\r\n",
      "\r\n",
      "def load_olga_model(model_directory, D=True):\r\n",
      "    \"\"\" Fully load an olga model.\r\n",
      "        model: directory where the model is stored\r\n",
      "        D: True if the model is VDJ (default)\r\n",
      "    \"\"\"\r\n",
      "    params_file_name = model_directory / 'model_params.txt'\r\n",
      "    marginals_file_name =  model_directory / 'model_marginals.txt'\r\n",
      "    V_anchor_pos_file =  model_directory / 'V_gene_CDR3_anchors.csv'\r\n",
      "    J_anchor_pos_file =  model_directory / 'J_gene_CDR3_anchors.csv'\r\n",
      "\r\n",
      "    if D:\r\n",
      "        genomic_data = olga.load_model.GenomicDataVDJ()\r\n",
      "        generative_model = olga.load_model.GenerativeModelVDJ()\r\n",
      "    else:\r\n",
      "        genomic_data = olga.load_model.GenomicDataVJ()\r\n",
      "        generative_model = olga.load_model.GenerativeModelVJ()\r\n",
      "        \r\n",
      "    genomic_data.load_igor_genomic_data(params_file_name, V_anchor_pos_file, J_anchor_pos_file)\r\n",
      "    generative_model.load_and_process_igor_model(marginals_file_name)\r\n",
      "    \r\n",
      "    if D:\r\n",
      "        pgen_model = olga.generation_probability.GenerationProbabilityVDJ(generative_model, genomic_data)\r\n",
      "    else:\r\n",
      "        pgen_model = olga.generation_probability.GenerationProbabilityVJ(generative_model, genomic_data)\r\n",
      "\r\n",
      "    return genomic_data, generative_model, pgen_model\r\n",
      "\r\n",
      "\r\n",
      "\r\n",
      "# load pgen model for TRA & TRB\r\n",
      "local_directory = Path(inspect.getfile(olga)).parent \r\n",
      "pgen = {}\r\n",
      "_, _, pgen['TRA'] = load_olga_model(\r\n",
      "    model_directory=local_directory/'default_models'/'human_T_alpha',\r\n",
      "    D=False)\r\n",
      "_, _, pgen['TRB'] = load_olga_model(\r\n",
      "    model_directory=local_directory/'default_models'/'human_T_beta',\r\n",
      "    D=True)\r\n",
      "\r\n",
      "\r\n",
      "\r\n",
      "def boundI1(r, meanS1, µp, σp, maxM=None):\r\n",
      "    \"\"\" return an upper bound for P[I(A, B) < r | A&B autologous] \"\"\"\r\n",
      "    λ = gmpy2.mpfr(meanS1) # parameter of the poisson law\r\n",
      "    if maxM is None:\r\n",
      "        maxM = 5 * meanS1\r\n",
      "    ks = range(1, maxM)\r\n",
      "    poiss = [λ**k/gmpy2.fac(k) * gmpy2.exp(-λ) for k in ks]\r\n",
      "    norms =  [1/2*(1 + gmpy2.erf((r - μp * k)/(gmpy2.sqrt(2 * k) * σp))) for k in ks]\r\n",
      "    if r <= 0:\r\n",
      "        return gmpy2.mpfr(0.)\r\n",
      "    else:\r\n",
      "        return (gmpy2.fsum([p * n for p, n in zip(poiss, norms)]) + \r\n",
      "                gmpy2.exp(-λ))\r\n",
      "\r\n",
      "\r\n",
      "def boundI2(r, M, meanS2, µp, σp, maxM=None):\r\n",
      "    \"\"\" return an upper bound for P[I(A, B) >= r | A&B not autologous] \"\"\"\r\n",
      "    λ = gmpy2.mpfr(meanS2) # parameter of the poisson law\r\n",
      "    if maxM is None:\r\n",
      "        maxM = 5*meanS2\r\n",
      "    ks = range(0, maxM)\r\n",
      "    poiss =  [gmpy2.exp(-λ)/gmpy2.fac(k) * λ**k for k in ks]\r\n",
      "    norms = [1/2*(1 + gmpy2.erf((r - μp * k)/(gmpy2.sqrt(2 * k) * σp))) \r\n",
      "             if k > 0 else 1 for k in ks]\r\n",
      "\r\n",
      "    if r > 0:\r\n",
      "        return gmpy2.fsum([p * (1-n) for p, n in zip(poiss, norms)])\r\n",
      "    else:\r\n",
      "        return gmpy2.mpfr(1.)\r\n",
      "\r\n",
      "\r\n",
      "def regularized_gamma_Q(n, mu):\r\n",
      "    return mpmath.gammainc(n, a=mu, regularized=True)\r\n",
      "\r\n",
      "    \r\n",
      "def rename_cols(df):\r\n",
      "    \"\"\"\r\n",
      "    Standardize the column names of the dataframe\r\n",
      "    \"\"\"\r\n",
      "    columns = df.columns.values\r\n",
      "    new_cols = []\r\n",
      "    for col in columns:\r\n",
      "        if col in translate_header:\r\n",
      "            new_cols.append(translate_header[col])\r\n",
      "        else:\r\n",
      "            new_cols.append(col)\r\n",
      "    df.columns = new_cols\r\n",
      "    return df\r\n",
      "\r\n",
      "\r\n",
      "def plot_S(m, ax, color=\"k\", label=None, alpha=1):\r\n",
      "    rg = np.arange(0, int(40*m))\r\n",
      "    ax.plot(rg, [scipy.stats.poisson.pmf(k, m) for k in rg], \r\n",
      "            color=color, # drawstyle='steps-post', \r\n",
      "            label=label, alpha=alpha,\r\n",
      "            linewidth=1)\r\n",
      "    return ax\r\n",
      "\r\n",
      "\r\n",
      "def plot_I(m, µ, σ, ax, label=None, alpha=1, color=\"k\"):\r\n",
      "    rg = np.linspace(0, 10, 100)\r\n",
      "    # precompute ? \r\n",
      "    Is = [sum(λ**k/fac(k) * np.exp(-λ) *\r\n",
      "          1/np.sqrt(2 * np.pi * n * σ**2) *\r\n",
      "          np.exp(-(x - μ*k)**2/(2 * σ**2 * k)\r\n",
      "          for k in range(0, 1000)) for x in rg)]\r\n",
      "    ax.plot(rg, Is,\r\n",
      "            color=color,\r\n",
      "            label=label, alpha=alpha,\r\n",
      "            linewidth=1)\r\n",
      "    return ax\r\n",
      "\r\n",
      "\r\n",
      "def estimate_S_from_counts(dfA, dfB):\r\n",
      "    cAs = dfA.groupby(\"cdr3\").read_count.sum().to_frame()\r\n",
      "    cBs = dfB.groupby(\"cdr3\").read_count.sum().to_frame()\r\n",
      "    cs = cAs.merge(cBs, left_index=True, right_index=True,\r\n",
      "                   how='outer', suffixes={'_A', '_B'})\r\n",
      "    cs[\"read_count\"] = cs[\"read_count_A\"] + cs[\"read_count_B\"]\r\n",
      "    M2 = len(cBs)\r\n",
      "    M1 = len(cAs)\r\n",
      "    # The formula below remove the bias associated with sampling\r\n",
      "    Sest = sum([1 + (gmpy2.bincoef(0, int(c)) \r\n",
      "                  - gmpy2.bincoef(M2, int(c))\r\n",
      "                  - gmpy2.bincoef(M1, int(c)))\r\n",
      "             /gmpy2.bincoef(M1+M2, int(c))  \r\n",
      "             for c in cs[\"read_count\"].values if c != 0])\r\n",
      "    return Sest\r\n",
      "    \r\n",
      "\r\n",
      "def immprint(dfA, dfB, full=False, use_counts=False, max_shared=np.inf):\r\n",
      "    \"\"\"\r\n",
      "    max_shared: If the number of sequences shared between dfA & dfB is\r\n",
      "    larger than max_shared, do not compute I.\r\n",
      "    \"\"\"\r\n",
      "    dfA = rename_cols(dfA)\r\n",
      "    dfB = rename_cols(dfB)\r\n",
      "\r\n",
      "    shared_sequences = set(dfA.cdr3_nt.values) & set(dfB.cdr3_nt.values)\r\n",
      "    S = len(set(dfA.cdr3_nt.values) & set(dfB.cdr3_nt.values)) \r\n",
      "    I = None\r\n",
      "    \r\n",
      "    if S < max_shared:\r\n",
      "        # if \"full\", the cdr3_nt column should have the standard form:\r\n",
      "        # TRA:..,TRB:...\r\n",
      "        I = -sum(np.log(compute_pgen(s, full)) for s in shared_sequences)\r\n",
      "    \r\n",
      "    # the mean of pgen is used to estimate I, it depends on\r\n",
      "    # the choice of model and on the scenario (homologous or\r\n",
      "    # not)\r\n",
      "    if full:\r\n",
      "        µ_logpgen_shared = 33\r\n",
      "        σ_logpgen_shared = 3\r\n",
      "        µ_logpgen = 47\r\n",
      "        σ_logpgen = 8\r\n",
      "    else: \r\n",
      "        µ_logpgen_shared = -18\r\n",
      "        σ_logpgen_shared = 2\r\n",
      "        µ_logpgen = -26\r\n",
      "        σ_logpgen = 6\r\n",
      "\r\n",
      "\r\n",
      "    µS2 = (1e-13 if full else 3.0e-07) * M1 * M2 #  default value (not autologous case)\r\n",
      "    if \"read_count\" in dfA.keys and \"read_count\" in dfB.keys() and use_counts:\r\n",
      "        µS1 = estimate_S_from_counts(dfA, dfB)\r\n",
      "    else:\r\n",
      "        µS1 = 7.96e-07 * M1 * M2  #  default values\r\n",
      "\r\n",
      "    return (S, I,\r\n",
      "            shared_sequences,\r\n",
      "            {\"µS1\": µS1, \"µS2\": µS2,\r\n",
      "             \"µ_logpgen_shared\": µ_logpgen_shared,\r\n",
      "             \"σ_logpgen_shared\": σ_logpgen_shared,\r\n",
      "             \"µ_logpgen\": µ_logpgen, \"σ_logpgen\": σ_logpgen})\r\n",
      "\r\n",
      "\r\n",
      "\r\n",
      "def parse_arguments():\r\n",
      "    parser = argparse.ArgumentParser(description='ImmPrint')\r\n",
      "    parser.add_argument('sampleA', type=str,\r\n",
      "                        help=\"csv file containing the sequences of the first sample\")\r\n",
      "    parser.add_argument('sampleB', type=str,\r\n",
      "                        help=\"csv file containing the sequences of the second sample\")\r\n",
      "    parser.add_argument('-f', '--full', dest='full',\r\n",
      "                        help=(\"If paired CDR3a / CDR3b are provided, use the full receptor \"\r\n",
      "                              \"to discriminate between the two scenarios. \"\r\n",
      "                              \"The CDR3 sequences should be provided in the column `cdr3_nt` (or `cdr3s_nt`)\"\r\n",
      "                              \"with the standard form: \\\"TRA:CDR3_TRA,TRB:CDR3_TRB\\\"\"),\r\n",
      "                        action='store_true')\r\n",
      "    parser.add_argument('-S', '--no_I_statistic', dest='S',\r\n",
      "                        help=(\"Don't compute the \\\"I\\\" statistics (potentially more precise but slower)\"),\r\n",
      "                        action='store_true')\r\n",
      "    parser.add_argument('-n', '--no-counts', dest='nocounts',\r\n",
      "                        help=\"If specified, don't use read counts, even if provided\",\r\n",
      "                        action='store_true')\r\n",
      "\r\n",
      "    args = parser.parse_args()\r\n",
      "\r\n",
      "    try:\r\n",
      "        dfA = pd.read_csv(args.sampleA, sep=None)\r\n",
      "    except IOError as e:\r\n",
      "        print(\"Error with opening the file corresponding to sample A.\\n\")\r\n",
      "        print(e)\r\n",
      "\r\n",
      "    try:\r\n",
      "        dfB = pd.read_csv(args.sampleB, sep=None)\r\n",
      "    except IOError as e:\r\n",
      "        print(\"Error with opening the file corresponding to sample B.\\n\")\r\n",
      "        print(e)\r\n",
      "\r\n",
      "    S, I, shared_sequences, parms = immprint(dfA, dfB, args.full, args.counts, args.S)\r\n",
      "\r\n",
      "    # Probability that the number of sequences shared is lower than S\r\n",
      "    # if both samples come from the same individual.\r\n",
      "    pv1 = regularized_gamma_Q(S+1, parms[\"µS1\"])\r\n",
      "    # Probability that the number of sequences shared is higher than S\r\n",
      "    # if both samples come from different individuals.\r\n",
      "    pv2 = 1 - regularized_gamma_Q(S+1, parms[\"µS2\"])\r\n",
      "\r\n",
      "\r\n",
      "    if not args[\"S\"]:\r\n",
      "        # Probability that a random Immprit, pgen-based, estimate is lower than I\r\n",
      "        # if both samples come from the same individual.\r\n",
      "        pvI1 = boundI1(I, µS1, parms[\"µ_logpgen\"], parms[\"σ_logpgen\"])\r\n",
      "        # Probability that a random Immprit, pgen-based, estimate is higher than I\r\n",
      "        # if both samples come from different individuals\r\n",
      "        pvI2 = boundI2(I, µS2, parms[\"µ_logpgen_shared\"], parms[\"σ_logpgen_shared\"])\r\n",
      "        if pvI1 < pv1:\r\n",
      "            pv1 = pvI1\r\n",
      "        if pvI2 < pv2:\r\n",
      "            pv2 = pvI2\r\n",
      "        \r\n",
      "    \r\n",
      "    \r\n",
      "    if pv1 < 1e-3: \r\n",
      "        print(f\"The two samples come from two different patients. \"\r\n",
      "              f\"The probability that they are autologous \"\r\n",
      "              f\"is lower than {pv1}\")\r\n",
      "    elif pv2 < 1e-3:\r\n",
      "        print(f\"The two samples come from the same patient. \"\r\n",
      "              f\"The probability that they were extracted in \"\r\n",
      "              f\"two different individuals is lower than {pv2}\")\r\n",
      "    else:\r\n",
      "        print(\"Ambiguous case: maybe not enough sequences.\")\r\n",
      "\r\n",
      "\r\n",
      "        \r\n",
      "    fig, ax = plt.subplots()\r\n",
      "    plot_S(µS1, ax, label=\"\")\r\n",
      "    ax.axvline(S)\r\n",
      "    plot_S(µS2, ax, label=\"XXX\")\r\n",
      "    ax.set_title(r\"$\\mathcal{S}$\")\r\n",
      "    plt.show()\r\n",
      "\r\n",
      "    fig, ax = plt.subplots()\r\n",
      "    plot_I(µS1, parms[\"µ_logpgen_shared\"], parms[\"σ_logpgen_shared\"], ax, label=\"XXX\")\r\n",
      "    ax.axvline(I)\r\n",
      "    plot_I(µS2, parms[\"µ_logpgen\"], parms[\"σ_logpgen\"], ax, label=\"XXX\")\r\n",
      "    ax.set_xlabel(r\"$\\mathcal{I}$\")\r\n",
      "    \r\n",
      "    plt.show()\r\n",
      "\r\n",
      "    \r\n",
      "    \r\n",
      "\r\n",
      "if __name__ == \"__main__\":\r\n",
      "    parse_arguments()\r\n"
     ]
    }
   ],
   "source": [
    "!cat /home/thomas/.local/lib/python3.8/site-packages/immprint/immprint.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
